src/include/mpir_gpu_util.h

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+#ifndef MPIR_GPU_UTIL_H_INCLUDED
+#define MPIR_GPU_UTIL_H_INCLUDED
+
+/* If buf is device memory, allocate a host buffer that can hold the data,
+ * return the host buffer, or NULL if buf is not device memory */
+MPL_STATIC_INLINE_PREFIX void *MPIR_gpu_host_alloc(const void *buf,
+                                                   MPI_Aint count, MPI_Datatype datatype)
+{
+    MPL_pointer_attr_t attr;
+    MPIR_GPU_query_pointer_attr(buf, &attr);
+
+    if (attr.type != MPL_GPU_POINTER_DEV) {
+        return NULL;
+    } else {
+        MPI_Aint extent, true_lb, true_extent;
+        MPIR_Datatype_get_extent_macro(datatype, extent);
+        MPIR_Type_get_true_extent_impl(datatype, &true_lb, &true_extent);
+
+        MPI_Aint shift = true_lb;
+        MPI_Aint size = true_extent;
+        if (count > 1) {
+            if (extent >= 0) {
+                size += extent * (count - 1);
+            } else {
+                MPI_Aint size_before = (-extent) * (count - 1);
+                /* extra counts are being packed *before*, so need *increase* the size */
+                size += size_before;
+                /* the allocated pointer will be *already* shifted by this amount */
+                shift -= size_before;
+            }
+        }
+        void *host_buf = MPL_malloc(size, MPL_MEM_OTHER);
+        MPIR_Assert(host_buf);
+
+        host_buf = (char *) host_buf - shift;
```

Comment:
So `true_lb` is always <= 0 when `count == 1` or `extent >= 0`? Am I reading this right?

Reply:
No. `true_lb` works whether it's positive or negative, it is an adjustment from the buffer pointer and the arithmetic will work for either positive or negative. However, the `true_lb` returned from `MPIR_Type_get_true_extent_impl` is the `true_lb` for a single count datatype. When `extent` is positive, the `true_lb` remain the same for multiple count.  so no adjustment is needed. However, when `extent` is negative, the `true_lb` with multiple count will be further extended into the negative direction, thus requires further adjustment.

Reply:
What I'm trying to understand is we are returning an allocated buffer, but with the address potentially shifted outside of the malloc'd region in the case of positive `true_lb`. Wouldn't writing to or dereferencing that address lead to a potential crash? How do we guard against these scenarios? Is it because `MPIR_Localcopy` will only be used with a datatype that would not do this?

Reply:
> Wouldn't writing to or dereferencing that address lead to a potential crash?

The dataype packing/unpacking routine should only access the actual data location specified by the MPI datatype. Only a bug will result in a wrong access.

Reply:
OK, only accessing this with a datatype-aware copy routine makes sense. I think it should be OK. I'd still like to see it go thru Jenkins again.

Reply:
The code has already passed Jenkins both here https://github.com/pmodels/mpich/pull/5176#issuecomment-809517249 and in the gpu testing https://github.com/pmodels/mpich/pull/5000#issuecomment-809516859 . The last push is only a rebase with PR #5005

Reply:
Per my previous comments, I am going to go ahead merge this, then get #5000 rebased. Any issues should pop up in that PR anyway.